src/buf_writer.c

```diff
@@ -134,3 +149,35 @@ buf_writer_terminate(tsdn_t *tsdn, buf_writer_t *buf_writer) {
 		buf_writer_free_internal_buf(tsdn, buf_writer->buf);
 	}
 }
+
+void
+buf_writer_set_allow_auto_flush(buf_writer_t *buf_writer,
+    bool allow_auto_flush) {
+	buf_writer_assert(buf_writer);
+	if (buf_writer->allow_auto_flush && !allow_auto_flush) {
```

Comment:
I slightly prefer only guarding with `allow_auto_flush`. The only case there's a difference is when manual is set multiple times, to ensure calling _commit() once for each _set() call, so the implicit revert don't go "out of scope".

Reply:
Agree that we should only guard with `allow_auto_flush`.

Upon a second thought, I think `buf_writer_set_allow_auto_flush(false)` shouldn't auto commit. Especially since we're changing the name `commit` to `checkpoint`, the caller should explicitly call `buf_writer_set_checkpoint()` every time they anticipate a potential rollback; otherwise the API shouldn't make any guarantee. This is mainly to force the caller side code to be clearer: it has to strictly follow a "three-step cycle" - `buf_writer_set_checkpoint()` -> `buf_writer_overflowed()` -> `handle_overflow()`.

On the buffered writer side, I'm always setting `checkpoint = 0` in `buf_writer_set_allow_auto_flush()`, just as a minimal safe default.

The other thing I was thinking: in addition to not doing auto commit, consecutive `buf_writer_set_allow_auto_flush(false)` calls should not auto clear the `overflowed` flag either. In other words, the `overflowed` flag can only be cleared in `buf_writer_set_allow_auto_flush(true)` or `buf_writer_set_checkpoint()`; the caller can thus freely check whether there has been an overflow at any time, even after flushing or after another `buf_writer_set_allow_auto_flush(false)` call, as long as it's before they set the next checkpoint. This is mainly to give the caller more flexibility in code structure, as long as they strictly follow the three-step cycle.

Overall, the `buf_writer_set_allow_auto_flush()` function should be more transparent: it should just set `allow_auto_flush`, as its name suggests, with only minimal / necessary additional operations.

Reply:
Agreed. Not auto setting sounds right.